Added support for create_attachment

T0ille opened the pull request:
🆕 Implemented create_attachment
Create one or multiple new attachments on a node

function arguments :
        :param project_id: ID for the project
        :param node_id: ID for the node to create attachment for
        :param attachment: Path of attachment to upload to a node
        :*args: path(s) of additional attachment(s) to add to the upload request (optional)

pathlib Path now required to read files (from pathlib import Path)

T0ille committed:
Update dradis.py

🆕 implemented create_attachment
Create one or multiple new attachment s on a node

function arguments :
        :param project_id: ID for the project
        :param node_id: ID for the node to create attachment for
        :param attachement: Path of attachement to upload to a node
        :*args: path(s) of additionnal attachement(s) to add to the upload request (optional)

pathlib now required to read files

T0ille committed:
Update README.md

removed the "attachments " line on unsupported endpoint  as it should now be supported

T0ille committed:
Update dradis.py

fixed typo

JJK96 commented on dradis.py:
```diff
@@ -965,25 +965,42 @@ def get_attachment(self, project_id: int, node_id: int, filename: int):
-        # Set the data
-        # data = {}
+        # Add file to upload
+        if attachment.is_file():
```
Why not just remove the whole "attachment" argument, since *args already covers the whole functionality? It could  be renamed to *attachments for clarity. It would remove some code duplication for the logic of adding files.

JJK96 commented on dradis.py:
```diff
@@ -965,25 +965,42 @@ def get_attachment(self, project_id: int, node_id: int, filename: int):
-    def create_attachment(self, project_id: int, node_id: int):
-        """Create a new attachment
+    def create_attachment(self, project_id: int, node_id: int, attachment: Path, *args: Path):
+        """Create new attachment() on a node
```
Extraneous ()

JJK96 commented on README.md:
```diff
@@ -60,4 +60,3 @@ Currently, the following endpoints are fully supported:
 The following endpoints are not fully supported:
 
 - The users endpoint is currently unsupported, feel free to implement support and submit a pull request.
-- The attachments endpoint is partially supported, the GET, UPDATE and DELETE operations are implemented, while the CREATE operation is not currently supported, since the endpoint uses the `multipart/form-data` content type instead of JSON. Feel free to implement support and submit a pull request.
```
If it's now fully supported, it can be added to the list above.

JJK96 commented:
Thank you for your merge request, nice concise implementation.

JJK96 commented:
I suggest you have a look at #2, I like how in that pull request a separate function is created for `_create_multipart`, this aids future implementations of other endpoints with multipart requests.

JJK96 commented:
@T0ille I see you resolved the conversations, but  I don't see any new commits. I unresolved them for now.

T0ille committed:
Dradis.py update

- Added verify_ssl option (true by default)
- added _create_multipart for multipart POST  requests
- simplified logic in create_attachment

T0ille commented:
@JJk96 I had forgotten to commit indeed 😅. Should be in commit [cb4dc87](https://github.com/NorthwaveSecurity/dradis-api/pull/1/commits/cb4dc873ba1524678530f7094bdfa1bde8e58097) now

> @T0ille I see you resolved the conversations, but I don't see any new commits. I unresolved them for now.

JJK96 commented:
Thank you for your contribution! I will review and probably merge it as soon as I find the time

T0ille committed:
Update dradis.py

Bug fix:
The _create_multipart would fail if there was no Content-type header on the multipart POST

JJK96 commented on dradis.py:
```diff
@@ -965,31 +965,55 @@ def get_attachment(self, project_id: int, node_id: int, filename: int):
         self._add_project_header(project_id)
+        
+        # Remove  application/json Content-type and let requests handle the multipart/format-data type + boundary automatically
+        self.__headers.pop('Content-type', None)
```
the __headers variable is mutated here, but it is used in other requests. So, other requests will not have the 'Content-Type' header set after _create_multipart is run. Make a new headers object with only the correct headers:

```
headers = {k:v for k,v in self.__headers.items() if k != 'Content-Type'}
```

T0ille replied:
Good point, thanks for the feedback.  I updated the code and took into account your comments.

JJK96 reviewed (changes requested):
2 comments from the last review were still open. And 1 new comment is added.

T0ille committed:
Minor updates

- Changed headers in _create_multipartdata
- Changed description for create_attachment
- Updated README  with supported features

JJK96 commented:
I only need to test the new functionality, then I will merge this.

JJK96 commented:
I see you blindly copy-pasted my out-of-the-head solution. I fixed the spelling error and merged it.

JJK96 commented:
Thank you for your contribution!
